ironfish/src/network/peerNetwork.ts

```diff
@@ -896,21 +963,49 @@ export class PeerNetwork {
       return false
     }
 
+    // TODO: We may want to remove this so that transactions still propagate
+    // even with a full worker pool
     if (this.node.workerPool.saturated) {
       return false
     }
 
-    const peersToSendTo = this.connectedPeersWithoutTransaction(hash)
+    return true
+  }
 
+  alreadyHaveTransaction(hash: TransactionHash): boolean {
     /*
-     * When we receive a new transaction we want to (1) sync it to the wallet (2) add it to the mempool
-     * and (3) send it to peers who haven't received it yet. If we've already done these 3 things then
-     * return early and don't validate and verify the transaction again */
-    if (
-      this.syncedTransactionHashes.has(hash) &&
-      this.node.memPool.exists(hash) &&
-      peersToSendTo.length === 0
-    ) {
+     * When we receive a new transaction we want to test if we have already processed it yet
+     * meaning we have it in the mempool or we have it on a block. */
+
+    let peersToSendTo = 0
+    for (const _ of this.connectedPeersWithoutTransaction(hash)) {
+      peersToSendTo++
+    }
+
+    return (
+      (this.node.memPool.exists(hash) || this.recentlyAddedToChain.has(hash)) &&
+      peersToSendTo === 0
+      // && TODO(daniel): also filter recently rejected (expired or invalid) transactions
+    )
+  }
+
+  private async onNewTransaction(peer: Peer, message: NewTransactionMessage): Promise<boolean> {
+    const received = new Date()
+
+    // Mark the peer as knowing about the transaction
+    const hash = new Transaction(message.transaction).hash()
+    peer.state.identity && this.markKnowsTransaction(hash, peer.state.identity)
+
+    // Let the fetcher know that a transaction was received and we no longer have to query it
+    this.transactionFetcher.receivedTransaction(hash)
+
+    if (!this.shouldProcessTransactions()) {
+      this.transactionFetcher.removeTransaction(hash)
```

Comment:
I wonder if we should make a `wrapper` function that calls receivedTransaction, onNewTransaction, then removeTransaction. I think having to remember to call removeTransaction before every return is a bit prone to error